instrumentation/spymemcached-2.12/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spymemcached/SpymemcachedTest.java

```diff
@@ -0,0 +1,985 @@
+import net.spy.memcached.CASValue;
+import net.spy.memcached.ConnectionFactory;
+import net.spy.memcached.ConnectionFactoryBuilder;
+import net.spy.memcached.DefaultConnectionFactory;
+import net.spy.memcached.MemcachedClient;
+import net.spy.memcached.internal.CheckedOperationTimeoutException;
+import net.spy.memcached.ops.Operation;
+import net.spy.memcached.ops.OperationQueueFactory;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testcontainers.containers.GenericContainer;
+
+class SpymemcachedTest {
+  private static final String KEY_PREFIX = "SpymemcachedTest-";
+  private static final int EXPIRATION = 3600;
+  private static final int TIMING_OUT_OPERATION_TIMEOUT = 1000;
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  @RegisterExtension static final AutoCleanupExtension cleanup = AutoCleanupExtension.create();
+
+  static GenericContainer<?> memcachedContainer;
+  static InetSocketAddress memcachedAddress;
+
+  @BeforeAll
+  static void setUp() {
+    memcachedContainer =
+        new GenericContainer<>("memcached:latest")
+            .withExposedPorts(11211)
+            .withStartupTimeout(Duration.ofMinutes(2));
+    memcachedContainer.start();
+    memcachedAddress =
+        new InetSocketAddress(
+            memcachedContainer.getHost(), memcachedContainer.getMappedPort(11211));
+  }
+
+  @AfterAll
+  static void cleanUp() {
+    memcachedContainer.stop();
+  }
+
+  private static MemcachedClient getMemcached() {
+    return getMemcached(builder -> {});
+  }
+
+  private static MemcachedClient getMemcached(Consumer<ConnectionFactoryBuilder> customizer) {
+    // Use direct executor service so our listeners finish in deterministic order
+    ExecutorService listenerExecutorService = MoreExecutors.newDirectExecutorService();
+
+    ConnectionFactoryBuilder connectionFactoryBuilder =
+        new ConnectionFactoryBuilder()
+            .setListenerExecutorService(listenerExecutorService)
+            .setProtocol(BINARY);
+    customizer.accept(connectionFactoryBuilder);
+
+    ConnectionFactory connectionFactory = connectionFactoryBuilder.build();
+    try {
+      MemcachedClient memcached =
+          new MemcachedClient(connectionFactory, Collections.singletonList(memcachedAddress));
+      cleanup.deferCleanup(memcached::shutdown);
+
+      // Add some keys to test on later:
+      Map<String, String> map = new HashMap<>();
+      map.put("test-get", "get test");
+      map.put("test-get-2", "get test 2");
+      map.put("test-append", "append test");
+      map.put("test-prepend", "prepend test");
+      map.put("test-delete", "delete test");
+      map.put("test-replace", "replace test");
+      map.put("test-touch", "touch test");
+      map.put("test-cas", "cas test");
+      map.put("test-decr", "200");
```

Comment:
Since we are calling this method at the start of every test, it would make the tests a little bit easier to reason about if `getMemcached()` had an argument for the map items to include for each scenario setup. That way the context around the starting values for each key is right there in each test. 

For example, this would help the reader understand things like: `assertThat(memcached.decr(key("test-decr"), 5)).isEqualTo(195)` without having to scroll back up to the top of the file.